mlir/tools/miopen-gen/miopen-gen.cpp

```diff
@@ -803,60 +822,342 @@ static func::FuncOp getMemsetFunc(ModuleOp module, mlir::Type elemType) {
+      // in_w_idx = out_w_idx * stride_w + fil_w_idx * dilation_w - padding_w_l;
+      heightIdx = b.create<AffineApplyOp>(
+          loc, heightMap,
+          ValueRange{ivs[6], ivs[genConfig.filterLayout.find('y')]});
+      widthIdx = b.create<AffineApplyOp>(
+          loc, widthMap,
+          ValueRange{ivs[7], ivs[genConfig.filterLayout.find('x')]});
+      break;
+    }
+
+    auto getIndices = [&](std::string tensor) {
+      SmallVector<mlir::Value, 5> result;
+      std::string layout;
+      if (tensor == "filter")
+        layout = genConfig.filterLayout;
+      else if (tensor == "input")
+        layout = genConfig.inputLayout;
+      else
+        layout = genConfig.outputLayout;
+      for (auto c : layout) {
+        auto direction = genConfig.operation.getValue();
+        if ((direction == miopen::ConvOpType::Fwd ||
+             direction == miopen::ConvOpType::BwdWeight) &&
+            tensor == "input") {
+          if (c == 'h') {
+            result.push_back(heightIdx);
+            continue;
+          } else if (c == 'w') {
+            result.push_back(widthIdx);
+            continue;
+          }
+        } else if (direction == miopen::ConvOpType::BwdData &&
+                   tensor == "output") {
+          if (c == 'h') {
+            result.push_back(heightIdx);
+            continue;
+          } else if (c == 'w') {
+            result.push_back(widthIdx);
+            continue;
+          }
+        }
+        result.push_back(ivs[loopIVs.find(c)]);
+      }
+      return result;
+    };
+
+    // Generate boundary testing
+    auto zeroIdx = b.create<arith::ConstantIndexOp>(loc, 0);
+    auto dimHeight = b.create<arith::ConstantIndexOp>(loc, dimH);
+    auto dimWidth = b.create<arith::ConstantIndexOp>(loc, dimW);
+
+    arith::CmpIOp testTop, testBottom, testLeft, testRight;
+    arith::AndIOp testHeight, testWidth, testAll;
+
+    // in_h_idx >= 0
+    testTop = b.create<arith::CmpIOp>(loc, mlir::arith::CmpIPredicate::uge,
+                                      heightIdx, zeroIdx);
+    // in_h_idx < in_h
+    testBottom = b.create<arith::CmpIOp>(loc, mlir::arith::CmpIPredicate::ult,
+                                         heightIdx, dimHeight);
+    // in_w_idx >= 0
+    testLeft = b.create<arith::CmpIOp>(loc, mlir::arith::CmpIPredicate::uge,
+                                       widthIdx, zeroIdx);
+    // i_w_idx < in_w
+    testRight = b.create<arith::CmpIOp>(loc, mlir::arith::CmpIPredicate::ult,
+                                        widthIdx, dimWidth);
+    testHeight = b.create<arith::AndIOp>(loc, testTop, testBottom);
+    testWidth = b.create<arith::AndIOp>(loc, testLeft, testRight);
+    testAll = b.create<arith::AndIOp>(loc, testHeight, testWidth);
+
+    if (genConfig.operation.getValue() == miopen::ConvOpType::BwdData) {
+      // out_h_tmp % stride_h == 0 && out_w_tmp % stride_w == 0
+      auto heightRemSIOp =
```

Comment:
Simalarly, can these be in the map?

Overall, one big map from for each tensor might be the right call